### PR TITLE
Place a button to create a list directly from "lists" page

### LINKED ISSUE #7390: Lists Improvements
It is disappointing that to create a list, one needs to navigate through many pages. Adding a list is not intuitive and many users may not create lists because they are unable to figure out how to do so. 

image

**### Describe the problem that you'd like solved**
Rectify issues with the creation of a user's LISTS.

1) Creating a reading list is not easy. It would be nice to be able to create a list directly from the " My Books" page.

Example:

image

2) The drop down menu " WANT TO READ"  where the user can create a list should be the be the same drop down menu on books returned from the book search.

Example:

When searching for a book, this is the drop down menu the user sees:

image

When clicking on a book, this is the drop down menu the user sees:

image

3)  " LISTS" should be named " My LISTS", to differentiate from public lists. 
Example of a public list.

image
 
These LISTS are also LISTS but they are not LISTS created by the user.

4)  " see all" should be switched to " edit" in all pages where user LIST appears. 

For example:

image

5) Switching " See all" to "[edit]" will allow users to click on edit to : rename a list, change the list description, delete lists.

6) Deleting all books from a list should not delete the list. Book and Lists should be separate so that if user wants to keep the list but delete all the books from the list and add new books, the user can. 
Currently, if the user wants to delete the last book on the list, the user sees:

image

**### Proposal & Constraints**
- [ ] #7613
2) Make all WANT TO READ drop down menus the same throughout the website. 
3) Rename LISTS to MY LISTS to differentiate user lists from public/ suggested lists.
4) Rename " See all" to [edit] in all pages where user MY LISTS appears. This will flow better with the site design and also provide users with a tool to edit their lists instead of just seeing user's lists.
5) Change " see all" to edit beside MY LISTS in all pages containing MY LISTS. User will click this button and be directed to a list of all their lists, The user can then change or delete any of their created lists.
6) Deleting all books from a user's list should not prompt the user to delete the entire list. User should have the option to keep an empty list.

**### Additional context**
These modifications are intended to enhance user engagement with Open Library and increase accessibility.

### PR DESCRIPTION
**Pull request by carlsonrob**
Closes #7390

Closes parts 1,3,4,5,6 of the ticket. Part 2 required an overhaul of the UI and degraded search performance.
<details>
<summary>1. Add 'create a new list' button to sidebar</summary>
<br>
Issue_1

Issue_1

</details>

<details>
<summary>3. Rename LISTS to MY LISTS to differentiate user lists from public/ suggested lists.</summary>
<br>
Issue_3

Issue_3

</details>

<details>
<summary>4/5. Rename " See all" to [edit] in all pages where user MY LISTS appears. </summary>
<br>
Issue_4-5

Issue_4-5

</details>

<details>
<summary>6. Retain list when last book is removed</summary>
<br>
Issue_6

Issue_6

</details>

### Technical

Currently, deleting a list isn’t updated in the sidebar until refresh. This also affects creating a list with the 'create a new list' button.

### Testing

- make css is required for the 'create a new list' button to appear

### Stakeholders

@DebbieSan @mekarpeles @cdrini

### CODE REVIEW

**Comment by mekarpeles** [comment]
Thank you for submitting this PR! I'll take a look through. We may want to stage some of these changes independently so pieces can be merged (since some may require design feedback). Will review shortly.

**Comment by cdrini** [comment]
Sorry for the confusion @carlsonrob ! We'll go over this on Friday to evaluate design tweaks :)

File: openlibrary/core/models.py [hunk 1]
```diff
@@ -860,22 +860,9 @@ def _get_lists_uncached(self, seed=None, limit=100, offset=0):
 
         return self._site.things(q)
 
-    def new_list(self, name, description, seeds, tags=None):
+    def new_list(self, name, description, tags=None):
```
Comment (cdrini) [inline openlibrary/core/models.py:863]: Let's roll back the changes in this file ; it turns out this API already supported creating a list without seeds. We don't want to make it only work without seeds!

File: openlibrary/plugins/openlibrary/js/lists/index.js [hunk 2]
```diff
@@ -209,11 +209,7 @@ function addCreateListClickListener(button, parentDropper) {
                 seeds: [seed],
             }
 
-            const successCallback = function(listKey, listTitle) {
```
Comment (cdrini) [inline openlibrary/plugins/openlibrary/js/lists/index.js:212]: Why was this removed?

File: openlibrary/templates/account/sidebar.html [hunk 3]
```diff
@@ -38,20 +38,25 @@
...
-              $# e.g. /people/openlibrary/lists/OL1L/MyList
-              $ path_id = ctx.path.split('/')[4]
-              $if list_id == path_id:
-                $ class_list = class_list + 'selected'
-            <li><a class="$class_list" data-ol-link-track="MyBooksSidebar|PatronList" href="/people/$username/lists/$list_id"><span class="li-count">$(len(lst.seeds))</span>$(lst['name'] if lst['name'] else '[%s]' % placeholder_name)</a></li>
-        </div>
+      $if owners_page:
+        <li class="section-header section-header-split"><span>$_('My Lists')</span> <a href="/people/$username/lists" data-ol-link-track="MyBooksSidebar|AllLists" class="li-count">$_('Edit') ($len(lists))</a></li>
+      $else:
+        <li class="section-header section-header-split"><span>$_('Lists')</span> <a href="/people/$username/lists" data-ol-link-track="MyBooksSidebar|AllLists" class="li-count">$_('See All') ($len(lists))</a></li>
```
Comment (cdrini) [inline openlibrary/templates/account/sidebar.html:44]: Collapse these into one; you can use `$(_('My lists') if owners_page else _('Lists))` without duplicating the wrapping html :)

File: static/css/components/mybooks-menu.less [hunk 4]
```diff
@@ -4,6 +4,59 @@
...
+.sidebar-section {
+  a.dropclick {
+    display: none;
+  }
+
+  .create {
+    text-align: center;
+  }
+
+  .create-new-list {
```
Comment (cdrini) [inline static/css/components/mybooks-menu.less:18]: I'm not seeing this anywhere in the html... is a piece missing perhaps?

**Review by cdrini — CHANGES_REQUESTED** [review]
Thanks @carlsonrob and @BrendenInhelder ! Some feedback for you :) I'm not seeing the "create a list" button at all; was that removed by accident?

File: openlibrary/core/models.py (hunk 1 above)
Comment (carlsonrob) [reply openlibrary/core/models.py:863]: Sure thing. What method creates a list without requiring a seed?

File: openlibrary/plugins/openlibrary/js/lists/index.js (hunk 2 above)
Comment (carlsonrob) [reply openlibrary/plugins/openlibrary/js/lists/index.js:212]: When the seed requirement to make a list was removed in lists.py:247 and models.py, the front-end still showed a book being added in the dropdown. I'll revert this so creating a list on the book's page still requires a seed.

File: static/css/components/mybooks-menu.less (hunk 4 above)
Comment (carlsonrob) [reply static/css/components/mybooks-menu.less:18]: The code:
$ user_lists = []
$:macros.ReadingLogDropper(user_lists)
was added to sidebar.html to create the alink with the classname “create-new-list”
This css targets this classname when it’s within the sidebar-section in sidebar.html

**Comment by carlsonrob** [comment]
> Thanks @carlsonrob and @BrendenInhelder ! Some feedback for you :) I'm not seeing the "create a list" button at all; was that removed by accident?

@cdrini 
Thanks for the feedback!
The ‘create a list’ button is from the alink in ReadingLogDropper.html:112
In sidebar.html, $:macros.ReadingLogDropper() generates an alink with the classname “create-new-list”.
The css update in mybooks-menu.less targets this classname when it’s within the sidebar-section.
It wasn’t clear how to isolate this list creation functionality in the sidebar and include its javascript synchronization and popup without going through this macro.

File: openlibrary/core/models.py (hunk 1 above)
Comment (cdrini) [reply openlibrary/core/models.py:863]: This one :P Specifying seeds as an empty list `[]` works as expected :+1:

Comment (carlsonrob) [reply openlibrary/core/models.py:863]: > This one :P Specifying seeds as an empty list `[]` works as expected 👍

Ok thanks.

File: static/css/components/mybooks-menu.less (hunk 4 above)
Comment (carlsonrob) [reply static/css/components/mybooks-menu.less:14]: This css was modified for the CreateList macro which has the button.

**Comment by DebbieSan** [comment]
This relates to: #7613

**Comment by DebbieSan** [comment]
Hi @carlsonrob, 

This is very close to ready.

I have attached the CSS for the button below if this makes things a bit easier design-wise:

/*For the box where the text goes*/

box-sizing: border-box;

position: absolute;
width: 106px;
height: 21px;
left: 491px;
top: 482px;

background: #EDEDED;
border: 1px solid #DDDDDD;
border-radius: 5px;


/* For the Create a list text*/

position: absolute;
width: 69px;
height: 15px;
left: 510px;
top: 485px;

/* Text */
font-family: 'Inter';
font-style: normal;
font-weight: 400;
font-size: 11.52px;
line-height: 14px;
display: flex;
align-items: center;

color: #666666;

In addition, are you able to change _edit_ to _see all >_ ? 
There should not be a distinction between logged-in and not logged-in users.  
It should be see all > all the time. 


Hope this helps! Sorry for the delay, I think with this PR linked to the issue it makes a bit easier to communicate. :)


TYSM

**Comment by carlsonrob** [comment]
@DebbieSan 
Here's how it looks now:
update

**Comment by carlsonrob** [comment]
@DebbieSan
Here's the update with the chevron after "See All":
chevron

File: openlibrary/macros/CreateList.html [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+                <div class="input">
+                    <button id="create-list-button" type="submit" class="larger">Create new list</button>
+                    &nbsp; &nbsp;
+                    <a class="small dialog--close plain red" href="javascript:;">Cancel</a>
+                </div>
+                <div class="label"></div>
+            </div>
+        </form>
+    </div>
+</div>
```
Comment (cdrini) [inline openlibrary/macros/CreateList.html:41]: This is identical to the modal as it appears from the drop down ; let's pull it out into its own template file to avoid duplicating this code.

File: openlibrary/macros/CreateList.html [hunk 6]
```diff
@@ -0,0 +1,41 @@
+<div class="widget-add">
```
Comment (cdrini) [inline openlibrary/macros/CreateList.html:1]: Have you tested this locally? I believe you will need this at the top of the file: `$def with (...)`

File: openlibrary/templates/account/sidebar.html [hunk 7]
```diff
@@ -38,20 +38,21 @@
...
-            $ list_id = lst.key.split('/')[-1]
-            $ class_list = ''
-            $if key == 'list':
-              $# e.g. /people/openlibrary/lists/OL1L/MyList
-              $ path_id = ctx.path.split('/')[4]
-              $if list_id == path_id:
-                $ class_list = class_list + 'selected'
-            <li><a class="$class_list" data-ol-link-track="MyBooksSidebar|PatronList" href="/people/$username/lists/$list_id"><span class="li-count">$(len(lst.seeds))</span>$(lst['name'] if lst['name'] else '[%s]' % placeholder_name)</a></li>
-        </div>
+        <li class="section-header section-header-split"><span>$(_('My lists') if owners_page else _('Lists'))</span> <a href="/people/$username/lists" data-ol-link-track="MyBooksSidebar|AllLists" class="li-count">$(_('See All'))<span class="chevron"></span></a></li>
```
Comment (cdrini) [inline openlibrary/templates/account/sidebar.html:41]: We don't need a separate element for this.

```suggestion
        <li class="section-header section-header-split"><span>$(_('My lists') if owners_page else _('Lists'))</span> <a href="/people/$username/lists" data-ol-link-track="MyBooksSidebar|AllLists" class="li-count chevron">$(_('See All'))</a></li>
```

File: openlibrary/templates/type/list/view_body.html [hunk 8]
```diff
@@ -85,17 +85,10 @@ <h1>$list.name</h1>
 
 window.q.push(function(){
     \$(".listDelete a").on('click', function() {
```
Comment (cdrini) [inline openlibrary/templates/type/list/view_body.html:87]: Could you expand what this is doing @carlsonrob ?

**Review by cdrini — CHANGES_REQUESTED** [review]
Thanks @carlsonrob ! This looks great, a few code changes to fix :)

File: openlibrary/macros/CreateList.html (hunk 5 above)
Comment (carlsonrob) [reply openlibrary/macros/CreateList.html:41]: I separated this out to be used by ReadingLogDropper and sidebar

File: openlibrary/macros/CreateList.html (hunk 6 above)
Comment (carlsonrob) [reply openlibrary/macros/CreateList.html]: I put this div in the sidebar rather than CreateList

File: openlibrary/templates/account/sidebar.html (hunk 7 above)
Comment (carlsonrob) [reply openlibrary/templates/account/sidebar.html]: I reflected this change in mybooks-menu.less as well

File: openlibrary/templates/type/list/view_body.html (hunk 8 above)
Comment (carlsonrob) [reply openlibrary/templates/type/list/view_body.html:87]: The list is being deleted when the last book is removed because of the original logic.
So #delete-list-dialog is being removed altogether here to prevent this.

Comment (carlsonrob) [reply openlibrary/templates/type/list/view_body.html:87]: There’s a merge conflict from changes to view_body.html recently #7765.
I’ll get your opinion here before reflecting the change in the other files.

Comment (kmcgonigle) [reply openlibrary/templates/type/list/view_body.html:87]: @carlsonrob that was my refactor that caused the merge conflict. The JS in `view_body.html` has been moved to its own independent file now, located at: [openlibrary/plugins/openlibrary/js/lists/ListViewBody.js](https://github.com/internetarchive/openlibrary/pull/7765/files#diff-f61b3cf03823c53ce981f0c1032c568509251e89bfb29fc7d95bcb52837d598c)

I didn't make any functional changes to the code (just added some code comments), so hopefully it will be somewhat straightforward to patch in your updates. 🤞 
I think the functionality you modified in this PR begins on line 59 of the new `ListViewBody.js` file.

Please let me know if you run into any issues resolving the merge conflict.

Comment (carlsonrob) [reply openlibrary/templates/type/list/view_body.html:87]: @kmcgonigle Thanks!

File: openlibrary/macros/CreateList.html
Comment (cdrini) [inline openlibrary/macros/CreateList.html:1]: Rename this to `CreateListModal`

File: openlibrary/macros/CreateList.html [hunk 9]
```diff
@@ -0,0 +1,34 @@
+<div class="hidden">
```
Comment (cdrini) [inline openlibrary/macros/CreateList.html:1]: Wrap this in a `$if render_once('CreateListModal'):`. We only need this HTML rendered once on a given page, since it can be re-used!

File: openlibrary/macros/ReadingLogDropper.html [hunk 10]
```diff
@@ -143,40 +143,7 @@ <h3>$_('Add to List')</h3>
     </div>
   </div>
   $if include_lists and render_once('lists/widget.addList'):
```
Comment (cdrini) [inline openlibrary/macros/ReadingLogDropper.html:145]: Then we can also remove this

```suggestion
  $if include_lists:
```

File: openlibrary/templates/account/sidebar.html [hunk 11]
```diff
@@ -38,20 +39,28 @@
...
-            $ list_id = lst.key.split('/')[-1]
-            $ class_list = ''
-            $if key == 'list':
-              $# e.g. /people/openlibrary/lists/OL1L/MyList
-              $ path_id = ctx.path.split('/')[4]
-              $if list_id == path_id:
-                $ class_list = class_list + 'selected'
-            <li><a class="$class_list" data-ol-link-track="MyBooksSidebar|PatronList" href="/people/$username/lists/$list_id"><span class="li-count">$(len(lst.seeds))</span>$(lst['name'] if lst['name'] else '[%s]' % placeholder_name)</a></li>
+        <li class="section-header section-header-split"><span>$(_('My lists') if owners_page else _('Lists'))</span> <a href="/people/$username/lists" data-ol-link-track="MyBooksSidebar|AllLists" class="li-count chevron">$(_('See All'))</a></li>
+        <div class="widget-add">
```
Comment (cdrini) [inline openlibrary/templates/account/sidebar.html:43]: Wrap this in a `$if owners_page:` ; we only want to show the create a list button for the owner!

File: static/css/components/mybooks-menu.less [hunk 12]
```diff
@@ -114,12 +133,24 @@
...
   }
+
+  /* stylelint-disable selector-max-specificity */
+  .li-count.chevron:after {
+    content: "\203A";
+    margin-left: 5px;
+    display: inline-block;
+    transform: scale(2.0);
+    position: absolute;
+    top: .2em;
```
Comment (cdrini) [inline static/css/components/mybooks-menu.less:151]: Small style tweaks. This makes the chevron a little lighter and better centered.

```suggestion
    margin: 0 5px;
    position: absolute;
    font-size: 1.5em;
    font-weight: normal;
    top: 50%;
    transform: translateY(-50%);
    right: 0;
    height: 100%;
```

File: static/css/components/mybooks-menu.less [hunk 13]
```diff
@@ -4,6 +4,25 @@
  *
  * http://localhost:8080/account/books/want-to-read
  */
+
+/* 'create a list' within sidebar */
+.sidebar-section {
+  .create {
+    text-align: center;
+  }
```
Comment (cdrini) [inline static/css/components/mybooks-menu.less:12]: We can remove the `.create` element entirely and style the `widget-add` instead.

```suggestion
  .widget-add {
    text-align: center;
    padding: 6px 0;
  }
```

**Review by cdrini — CHANGES_REQUESTED** [review]
Thanks @carlsonrob ! A few coding fixes :)

Also noticed a UX issue; after creating the list, nothing happens which is a bit confusing. Can you add a on submit listener to the form? And make it so that when the form is submitted the page reloads? Ideally it would go to the list, but getting that information is a little complicated, so just reloading will do for now. Note this should only happen from the my books page, not the other place the create list modal exists.

Also a list page without any books on it is a bit of a confusing environment! Can you make it so that if a list is empty, it displays: "There are no items on this list." for non-owners, and "There are no items on this list! <a href="/search">Search for book, subjects, or authors</a> to add to your list. <a href="https://openlibrary.org/help/faq/reading-log#lists">Learn more.</a>"

**Comment by cdrini** [comment]
Thank you so much for your patience, @carlsonrob ! I know this one is taking a while to work through the pipe :)

**Comment by carlsonrob** [comment]
@cdrini 
> Also noticed a UX issue; after creating the list, nothing happens which is a bit confusing. Can you add a on submit listener to the form? And make it so that when the form is submitted the page reloads?

I added a form listener to sidebar.html so the page reloads on submission e356e2202317b3c12c165c954f6563ebd0241e7c. In the technical section of this PR, I mentioned this UX issue also occurs when deleting a list. Should an issue be created to resolve both of these cases more thoroughly?

> Also a list page without any books on it is a bit of a confusing environment! Can you make it so that if a list is empty, it displays: "There are no items on this list." for non-owners, and "There are no items on this list! [Search for book, subjects, or authors](/search) to add to your list. [Learn more.](https://openlibrary.org/help/faq/reading-log#lists)"

I updated lists.html to display those messages when the list is empty 2b4236628f91d24b61d1c86abde7bef11125ac60. When the list becomes empty, a refresh is still required to display these messages.

File: openlibrary/templates/type/list/view_body.html (hunk 8 above)
Comment (carlsonrob) [reply openlibrary/templates/type/list/view_body.html]: @cdrini This change was reflected here 0c2cc52bf7205a00acb2f160245a447fce011175. It's to keep the list after deleting the last item. For example, preventing this popup: 
remove_seed_dialog

**Review by carlsonrob — COMMENTED** [review]
Comments inline

File: openlibrary/plugins/openlibrary/js/lists/ListViewBody.js [hunk 14]
```diff
@@ -55,8 +83,9 @@ const getConfirmButtonLabelText = () => {
 };
 
 // Add listeners to each .listDelete link element
-$('.listDelete a').on('click', function() {
-    if (get_seed_count() > 1) {
+// Sometimes .listDelete is dynamically added to the DOM, so we'll add the listener to a parent element
```
Comment (cdrini) [inline openlibrary/plugins/openlibrary/js/lists/ListViewBody.js:86]: Why is it that .listDelete is added dynamically to the DOM? It should always be rendered by the server?

**Review by cdrini — COMMENTED** [review]
Hi @carlsonrob ! I made a few changes to your MR (see the last few commits). There were some downstream changes that make this UX much simpler! There is now a dedicate page for creating a list, so I made the button just link to it :+1:

Otherwise, I have one question about some of these JS changes ; I'm not sure why we need them? The delete links should always already be present, no?

Also I found a but where it says `See All` in the side bar even if the user is the owner.

Otherwise I think those might be the last changes!

File: openlibrary/plugins/openlibrary/js/lists/ListViewBody.js (hunk 14 above)
Comment (carlsonrob) [reply openlibrary/plugins/openlibrary/js/lists/ListViewBody.js:86]: Originally the change to ListViewBody was to remove the (get_seed_count > 1) statement so that removing the last item from a list didn't prompt the user to delete the whole list. 0c2cc52bf7205a00acb2f160245a447fce011175

The intent of commit 0abc0ad74bd572b714898488a76b4c7ea5c8a2af was to rollback those changes as that statement was recently updated and affirmed. It seems like that rollback didn't go through all the way. Lets make it so ListViewBody.js isn't updated any longer. Thanks!

**Comment by carlsonrob** [comment]
@cdrini
> Also I found a bug where it says `See All` in the side bar even if the user is the owner.

Currently when logged in, I see "SEE ALL" on production vs "SEE ALL" with the chevron on the test environment. Is this not expected?
Thanks.

seeAll

**Review by cdrini — APPROVED** [review]
Tested:
- Clicking "create a new list" sends you to the add page on my books page
- Clicking add to new list on book page shows popup.

Lgtm!
